### PR TITLE
feature: Introduce `YieldFromArrayToYieldsFixer`

### PR DESCRIPTION
**Pull request by kubawerlos**
Didn't spend much time on the name and description.

See the diff for tests in this repo at https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7115.

### CODE REVIEW

File: doc/list.rst [hunk 1]
```diff
@@ -3265,6 +3265,11 @@ List of Available Rules
    Part of rule sets `@PhpCsFixer <./ruleSets/PhpCsFixer.rst>`_ `@Symfony <./ruleSets/Symfony.rst>`_
 
    `Source PhpCsFixer\\Fixer\\ArrayNotation\\WhitespaceAfterCommaInArrayFixer <./../src/Fixer/ArrayNotation/WhitespaceAfterCommaInArrayFixer.php>`_
+-  `yield_from_array_to_yields <./rules/array_notation/yield_from_array_to_yields.rst>`_
+
+   Yield from array should be unpacked to series of yields.
```
Comment (keradus) [inline doc/list.rst:3270]: would be good to add reasoning for JohnDoe to understand benefit of using this rule

**Comment by Wirone** [comment]
@kubawerlos How hard would it be to provide 2-way conversion? Maybe there are people, who prefer single `yield from` and would be happy if they could compact multiple `yield`s automatically 🙂.

**Comment by keradus** [comment]
I would suggest to have good reasoning behind the rule, and have a way follow the good reasoning and not the unfortunate reasoning. eg if one way is, dunno, more performing to execute, we should promote only that way - imho

**Comment by Wirone** [comment]
I am not aware of any real advantage, I think it's just a matter of taste, like `?int` vs `int|null` 🙂. Correct me if I'm wrong.

**Comment by kubawerlos** [comment]
> I think it's just a matter of taste

Isn't quite non-subjective that having 1 less dimension in an array makes it easier to read?

`yield from` 3-dimensional array (quite often in PHPUnit's data providers, when a test has an array in arguments) is converted to yields of 2-dimensional arrays.

Also, let's not forget about the case we already have in few places in tests:
```php
function provideFixCases()
{
    yield from [
        // many lines
    ];

    yield 'one more test' => [/* ... */];

    yield 'one yet more test' => [/* ... */];
}
```

**Comment by Wirone** [comment]
At this point I am pretty sure that when it comes to "readability" there's no objective truth 😅. I saw so many opinions that code X is "more readable" that I don't agree with, that it does not surprise me anymore.

In terms of your example, it could go either way - expand `yield from` to multiple `yield` or include hanging `yield` into group above. What's more readable is subjective imho, but personally I prefer single `yield` since for me it's easier to catch start/end of each element (in multi-dimensional array it can be harder).

**Comment by kubawerlos** [comment]
> At this point I am pretty sure that when it comes to "readability" there's no objective truth sweat_smile.

Of course when are talking about `is_null` vs `=== null`, but there is some - if you remove all the `\n` from a file no one will claim it will be more readable and I'd advocate the same about arrays - less dimensions means more readable.

Have you (or anyone else) heard of a desire of packing anything into `yield from`? And more: for some good reason?

My motivation for the fixer were mostly the test cases we have here. We migrated array multiple times to yields (or more often to `yield from` followed by `yields` for new cases), never the other way. With this fixer anyone can easily put `yield from` and the fixer unpack it nicely.

**Comment by Wirone** [comment]
I agree in general, because as I said I also prefer unpacked yields. I just asked if it would be hard to add support for migrating the other way 😆.

**Comment by keradus** [comment]
it's not about "is it hard". it's about "is it recommended" - and allow that only if we do.
if we have 3 ppl so far up for promoting "unpacking" `yield from`, but nobody promoting alternative - I believe we do not need to offer it. Let's offer it only if there is aligned value to use the alternative.

prime rule: we should allow for best practices and not for everything that is technically possible
secondary rule: allowing for everything make overall code more complex (doesn't matter if much complex or bit complex)

**Comment by Wirone** [comment]
"Best practices" are always (often?) subjective, and current fixers are not consistent with what you're telling now: there's for example `array_syntax`, not `short_array_syntax`, even though it's rather discouraged to use `array()` now.

Having ability to change code both ways allows making upgrade/downgrade rule sets (like in Rector), while fixers with 1-way conversion can't be used for downgrading (which is an awesome approach for writing modern code but providing it to end users with lower PHP version compatibility - like PHPStan is doing, where source code is written in 8.1 but in the end tool is compatible with 7.2). So IMHO it's not always about "best practices", but about automating changes. Default config may represent "best practices".

But it's only digression anyway 😅. I believe `yield` and `yield from` were introduced in the same version of PHP (docs are not clear about it), so it does not relate to upgrading/downgrading. I just wanted to share my perspective, that we shouldn't focus on best practices on fixers' features level, rather on fixers' default config level.

**Comment by keradus** [comment]
> "Best practices" are always (often?) subjective,

I believe we should discuss and promote best practices, if sth is subjective, we should come with a reasoning, as I asked in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7114#discussion_r1253246209

> and current fixers are not consistent with what you're telling now: there's for example array_syntax, not short_array_syntax, even though it's rather discouraged to use array() now.

I'm very happy to have this simplified and have not so good rules deprecated

> Having ability to change code both ways allows making upgrade/downgrade rule sets (like in Rector), while fixers with 1-way conversion can't be used for downgrading (which is an awesome approach for writing modern code but providing it to end users with lower PHP version compatibility - like PHPStan is doing, where source code is written in 8.1 but in the end tool is compatible with 7.2). So IMHO it's not always about "best practices", but about automating changes. Default config may represent "best practices".

I want this project to promote best practices and while I see reason to keep compatibility with older versions of PHP, I want to keep it sane. If less that 10% of users using old version of PHP, I do not see the need to provide cutting-edge version of PHP CS Fixer for them. They can still use older version of PHP CS Fixer and put focus and effort to migrate their codebase to modern PHP rather than use cutting-edge PHP CS Fixer, yet legacy PHP.
With that, I do not aim with PHP CS Fixer to provide (even auto-downgraded) support for legacy PHP versions, and I'm OK to not use cutting-edge PHP 8.3 for development of tool, while we still support 8.0.

And hey, one can always run Fixer via docker.

> that we shouldn't focus on best practices on fixers' features level, rather on fixers' default config level.

every "multiple possibilities" is bringing more complexity, if we don't recommend something and we do not hear need for that, let's not create/maintain it.

File: tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                    yield "bar" => function ($x, $y) { return max($x, $y); };
+                '.'
+            }',
+            '<?php function f() {
+                yield from [
+                    random_key() => true,
+                    "foo" => foo(1, 2),
+                    "bar" => function ($x, $y) { return max($x, $y); },
+                ];
+            }',
```
Comment (keradus) [inline tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php:114]:
```suggestion
            '<?php function f() {
                 '.'
                    yield random_key() => true;
                    yield "foo" => foo(1, 2);
                    yield "bar" => function ($x, $y) { return max($x, $y); };
                    yield "baz" => function () { yield 123; };
                '.'
            }',
            '<?php function f() {
                yield from [
                    random_key() => true,
                    "foo" => foo(1, 2),
                    "bar" => function ($x, $y) { return max($x, $y); },
                    "baz" => function () { yield 123; },
                ];
            }',
```

File: tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                function f5() {  yield 8; yield 9; }
+            ',
+            '<?php
+                function f1() { yield from array(0, 1); }
+                function f2() { yield from [2, 3]; }
+                function f3() { yield from array(4, 5); }
+                function f4() { yield from [6, 7]; }
+                function f5() { yield from array(8, 9); }
+            ',
+        ];
```
Comment (keradus) [inline tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php:145]:
```php
function foo() {
    yield from [
        "this element is regular string",
        yield from ["here", "are", "nested", "strings"],
        "next element will be an arrow function reference",
        fn() => [yield 1, yield from [2, 3]]
    ];
}
```

can we add sth like this? I didn't found similar case already present

File: src/Fixer/ArrayNotation/YieldFromArrayToYieldsFixer.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+            if (!$tokens[$arrayStartIndex]->isGivenKind([T_ARRAY, CT::T_ARRAY_SQUARE_BRACE_OPEN])) {
+                continue;
+            }
+
+            $tokens->clearTokenAndMergeSurroundingWhitespace($index);
+
+            if ($tokens[$arrayStartIndex]->isGivenKind(T_ARRAY)) {
+                $startIndex = $tokens->getNextTokenOfKind($arrayStartIndex, ['(']);
+                $tokens->clearAt($arrayStartIndex);
```
Comment (keradus) [inline src/Fixer/ArrayNotation/YieldFromArrayToYieldsFixer.php:84]: I found small bug for this line, fix (and few other changes) in https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/1/commits/26de244b9c8176ccf8885ca2040b348e7410b251

File: src/Fixer/ArrayNotation/YieldFromArrayToYieldsFixer.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+/**
+ * @author Kuba Werłos <werlos@gmail.com>
+ */
+final class YieldFromArrayToYieldsFixer extends AbstractFixer
+{
+    /**
+     * @var array<int, Token>
+     */
+    private array $inserts;
```
Comment (keradus) [inline src/Fixer/ArrayNotation/YieldFromArrayToYieldsFixer.php:33]: myself, I am trying to keep Fixers stateless, ie not having interim states in properties (ideally all properties are readonly and injected via constructor (yeah, we don't have DIC yet)).

Here, we populate this property in one method, consume in another, JohnDoe need to be mindful to not break interim state of this property and call functions in proper order.

I would suggest to change it - https://github.com/6b7562617765726c6f73/PHP-CS-Fixer/pull/2/commits/a49645015b1a8fb217b7a049cc5da0a5fc6fef9f

File: tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+        yield [
+            '<?php
+                function foo() {
+                     '.'
+                        yield "this element is regular string";
+                        yield yield from ["here", "are", "nested", "strings"];
```
Comment (keradus) [inline tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php:174]: I know I was coming with the code variant in `before fixing` version here, but is `yield yield from ...` is valid and meaningful syntax?
would it be same as `yield from` ? should we unpack?

File: tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                        yield "next elements will be an arrow function reference";
+                        yield fn() => [yield 1, yield from [2, 3]];
+                        yield fn() => [yield from [1, 2], yield 3];
+                        yield fn() => [yield from array(1, 2), yield 3]
```
Comment (keradus) [inline tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php:178]: should we fix nested `yield from` ?

File: tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php (hunk 6 above)
Comment (kubawerlos) [reply tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php]: Using "return value" of `yield from` does not make any sense in the first place.

File: tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php (hunk 7 above)
Comment (kubawerlos) [reply tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php]: I don't think so, either doing `yield yield from` or `yield from` from the nested `yield from` will change the behavior: https://3v4l.org/CjdTH - better not touch `yield from` having nested `yield from` at all.

That must be a new record for using word "yield" in single sentence 😁

Comment (keradus) [reply tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php]: I believe nested yields, when behind callback, are independent and should be fixed.

pls check this out
https://3v4l.org/tck3Q

Comment (kubawerlos) [reply tests/Fixer/ArrayNotation/YieldFromArrayToYieldsFixerTest.php]: There is a different value if you call it this way: https://3v4l.org/mLt9f

**Review by keradus — APPROVED** [review]

**Pull request merged**